Feat: Implement unlock_funds_with_proof with zkProof validation

Guzbyte-tech opened the pull request:
## Summary

This PR integrates Starknet's Solidity Verifier (GpsStatementVerifier) into the ZeroXBridge smart contract, allowing the verification of L2 zkProofs and unlocking ERC-20 tokens on L1. It ensures seamless cross-layer transactions by validating STARK proofs via verifyProofAndRegister.

## Implementation Details

# GpsStatementVerifier Integration:
- Implements IGpsStatementVerifier interface to call verifyProofAndRegister.
- Uses GpsStatementVerifier to validate zkProofs from SHARP.

# Proof Verification & Security Enhancements:
- Replay Attack Prevention: Tracks used proofs & commitment hashes to prevent reuse.
- Commitment Hash Validation: Ensures proofs correspond to expected user data.

# Ownership & Access Control:
- Relayer Approval System: Only approved relayers can submit proofs.
- Owner-Restricted Configuration: GPS verifier & Cairo program ID can be updated by the contract owner.

# ERC-20 Token Handling:
- Unlocked funds are stored as claimable balances.
- Users can claim funds via claimFunds().

# Tests Implemented
- Proof Verification: Ensures valid proofs unlock funds.
- Failure Cases: Invalid proofs, reused proofs, incorrect commitment hashes.
- Ownership & Relayer Controls: Tests for onlyOwner functions & relayer access.
- Token Transfers: Users can successfully claim unlocked funds.

This follows Starknet’s modular verifier structure, where GpsStatementVerifier serves as the primary verification contract. The approach ensures security, efficiency, and compatibility with SHARP proofs.

TEST

<img width="904" alt="image" src="https://github.com/user-attachments/assets/83bff4f7-7ec9-4b84-8345-5c1c90f909dd" />

Close #2

JoE11-y commented:
HI @Guzbyte-tech great work man.

Just a few things, can you rename the file to be ZeroXBridgeL1.sol and fix the merge conflicts.

Guzbyte-tech commented:
Hi @JoE11-y, I have:
-  Renamed and modified ZeroXBridgeL1.sol to suit the starkness verifier.
-  Modified the ZeroXBridgeL1.t.sol to suit the constructor argument in the modified ZeroXBridgeL1.sol contract.
-  Moved the unluck_funds_proof with zkProof validation test suites to a new file `ZeroXBridgeProofAndRelayerTest.t.sol`.
- Fixed merge conflicts
- All existing test passing

<img width="1234" alt="image" src="https://github.com/user-attachments/assets/93d23d1d-d003-4c75-8c35-6567e26c89c6" />

JoE11-y commented:
LGTM!
